### PR TITLE
New: print proto error details if any in the returned error

### PR DESCRIPTION
**Pull request by Dogild**
Error details are now printed along the protobuf error if there are any.

For example you will get the following:
```
Failed while making call: code:internal message:this is an expected error
{
  "details": [
    {
      "type.googleapis.com/uber.infra.net.rpc.probe.EchoErrorDetail": {
        "payload": {
          "pongInt32": 1
        }
      }
    }
  ]
}
```

If no error details are returned, the output will remain the same as before, for instance:
```
Failed while making call: code:internal message:this is an expected error
```

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,9 @@
 Changelog
 =========
 
+# Unreleased
+* New: gRPC details are now printed along the error if there are any.
```
Comment (AllenLuUber) [inline CHANGELOG.md:5]: if there is any

File: encoding/protobuf.go [hunk 2]
```diff
@@ -106,6 +108,50 @@ func (p protoSerializer) Encoding() Encoding {
 	return Protobuf
 }
 
+func (p protoSerializer) ErrorDetails(err error) ([]interface{}, error) {
+	// Here we use yarpcerrors since the transport layer of yab is using yarpc as well
+	if err == nil || !yarpcerrors.IsStatus(err) {
```
Comment (AllenLuUber) [inline encoding/protobuf.go:113]: no need to check `err == nil`: 
from [IsStatus doc](https://pkg.go.dev/go.uber.org/yarpc@v1.53.2/yarpcerrors#IsStatus): This is false if the error is nil.

File: encoding/protobuf.go [hunk 3]
```diff
@@ -106,6 +108,50 @@ func (p protoSerializer) Encoding() Encoding {
...
+		return nil, nil
+	}
+
+	details := []interface{}{}
```
Comment (AllenLuUber) [inline encoding/protobuf.go:117]: can we move this closer to where it is actually getting used? say before L128

File: encoding/protobuf.go [hunk 4]
```diff
@@ -106,6 +108,50 @@ func (p protoSerializer) Encoding() Encoding {
...
+	yerr := yarpcerrors.FromError(err)
+	if len(yerr.Details()) == 0 {
+		return nil, nil
+	}
+
+	errStatus := &status.Status{}
+	if err := proto.Unmarshal(yerr.Details(), errStatus); err != nil {
+		return nil, fmt.Errorf("could not unmarshal error details %s", err.Error())
```
Comment (AllenLuUber) [inline encoding/protobuf.go:125]: do you want to wrap error with %w?
`fmt.Errorf("could not unmarshal error details %w", err)`
https://blog.golang.org/go1.13-errors

File: encoding/protobuf.go [hunk 5]
```diff
@@ -106,6 +108,50 @@ func (p protoSerializer) Encoding() Encoding {
...
+	}
+
+	for _, detail := range errStatus.Details {
+		rdetail, rerr := p.anyResolver.Resolve(detail.TypeUrl)
+		if rerr != nil {
+			// If we reach this point it means we are not able to resolve the type of the detail
+			// This can happen when an error is being bubbled up in a chaine of services.
```
Comment (AllenLuUber) [inline encoding/protobuf.go:132]: chain

File: encoding/protobuf.go [hunk 6]
```diff
@@ -106,6 +108,50 @@ func (p protoSerializer) Encoding() Encoding {
...
+		}
+
+		if err := proto.Unmarshal(detail.Value, rdetail); err != nil {
+			return nil, fmt.Errorf("could not unmarshal error detail message %s", err.Error())
+		}
+
+		details = append(details, map[string]interface{}{
+			detail.TypeUrl: rdetail,
+		})
+	}
```
Comment (AllenLuUber) [inline encoding/protobuf.go:150]: seeing 2 append here. Can it be simplifed to :
```go
	for _, detail := range errStatus.Details {
		rdetail, rerr := p.anyResolver.Resolve(detail.TypeUrl)
                 // comments on why the default value
		var val interface{} = detail.Value
		
		if rerr == nil {
			if err := proto.Unmarshal(detail.Value, rdetail); err != nil {
				return nil, fmt.Errorf("could not unmarshal error detail message %s", err.Error())
			}
		}

		details = append(details, map[string]interface{}{
			detail.TypeUrl: rdetail,
		})
	}
```

File: main.go [hunk 7]
```diff
@@ -472,7 +473,27 @@ func makeContextWithTrace(ctx context.Context, t transport.Transport, request *t
 func makeInitialRequest(out output, transport transport.Transport, serializer encoding.Serializer, req *transport.Request) {
 	response, err := makeRequestWithTracePriority(transport, req, 1)
 	if err != nil {
-		out.Fatalf("Failed while making call: %v\n", err)
+		buffer := bytes.NewBuffer(nil)
+		_, _ = buffer.WriteString(err.Error())
```
Comment (AllenLuUber) [inline main.go:477]: `buffer.WriteString(err.Error())`. Same for all the last 3 `WriteString`

File: encoding/protobuf.go (hunk 4 above)
Comment (Dogild) [reply encoding/protobuf.go:124]: we never used %w elsewhere in this project. I set %s to stay consistent with the other errors.

File: main.go [hunk 8]
```diff
@@ -472,7 +473,27 @@ func makeContextWithTrace(ctx context.Context, t transport.Transport, request *t
 func makeInitialRequest(out output, transport transport.Transport, serializer encoding.Serializer, req *transport.Request) {
 	response, err := makeRequestWithTracePriority(transport, req, 1)
 	if err != nil {
-		out.Fatalf("Failed while making call: %v\n", err)
+		buffer := bytes.NewBuffer(nil)
+		buffer.WriteString(err.Error())
```
Comment (AllenLuUber) [inline main.go:477]: https://golang.org/pkg/bytes/#NewBufferString

**Review by AllenLuUber — APPROVED** [review]

File: main.go [hunk 9]
```diff
@@ -472,7 +473,27 @@ func makeContextWithTrace(ctx context.Context, t transport.Transport, request *t
...
+				out.Fatalf("Failed to get protobuf error details %s", derr.Error())
+			}
+
+			if len(details) > 0 {
+				bs, merr := json.MarshalIndent(map[string]interface{}{"details": details}, "", "  ")
+				if merr != nil {
+					out.Fatalf("Failed to convert protobuf error details to JSON: %v\nMap: %+v\n", merr, details)
+				}
+				buffer.WriteString("\n")
+				buffer.WriteString(string(bs))
```
Comment (AllenLuUber) [inline main.go:491]: `buffer.Write(bs)`

File: encoding/protobuf_test.go [hunk 10]
```diff
@@ -159,6 +161,138 @@ func TestProtobufRequest(t *testing.T) {
...
+		t.Run(tt.desc, func(t *testing.T) {
+			serializer, err := NewProtobuf(tt.method, tt.source)
+			require.NoError(t, err, "Failed to create serializer")
+			errorSerializer, ok := serializer.(ProtoErrorSerializer)
+			require.True(t, ok)
+
+			result, err := errorSerializer.ErrorDetails(tt.err)
+
+			if tt.wantErr != nil {
+				require.Error(t, err, "%v", tt.desc)
```
Comment (AllenLuUber) [inline encoding/protobuf_test.go:284]: require.Error(t, err, tt.desc)

File: encoding/protobuf_test.go [hunk 11]
```diff
@@ -159,6 +161,138 @@ func TestProtobufRequest(t *testing.T) {
...
+		wantOutAsJSON string
+		wantErr       error
+	}{
+		{
+			desc:          "nil error",
+			source:        source,
+			method:        "Bar/Baz",
+			err:           nil,
+			wantErr:       nil,
+			wantOutAsJSON: `null`,
```
Comment (AllenLuUber) [inline encoding/protobuf_test.go:216]: "null", same as the others

File: encoding/protobuf.go (hunk 4 above)
Comment (AllenLuUber) [reply encoding/protobuf.go:124]: ok I can buy this

File: encoding/encoding.go [hunk 12]
```diff
@@ -69,6 +69,13 @@ type Serializer interface {
 	MethodType() MethodType
 }
 
+// ProtoErrorSerializer deserializes errors.
+// Error details are very specific to proto.
+type ProtoErrorSerializer interface {
```
Comment (jronak) [inline encoding/encoding.go:74]: Nit: should we name this interface `ProtoErrorDetailsDecoder` and its method as `Decode`?

File: encoding/protobuf.go [hunk 13]
```diff
@@ -106,6 +108,50 @@ func (p protoSerializer) Encoding() Encoding {
...
+
+	errStatus := &status.Status{}
+	if err := proto.Unmarshal(yerr.Details(), errStatus); err != nil {
+		return nil, fmt.Errorf("could not unmarshal error details %s", err.Error())
+	}
+
+	details := []interface{}{}
+	for _, detail := range errStatus.Details {
+
```
Comment (jronak) [inline encoding/protobuf.go:129]: nit: remove extra line

**Review by jronak — APPROVED** [review]

File: encoding/encoding.go (hunk 12 above)
Comment (Dogild) [reply encoding/encoding.go]: I went with `ProtoErrorDeserializer` to stay consistent with the other interface `Serializer`

**Pull request merged**